Cory Adams Moulton (V2)

cmikida2 opened the pull request:
This pull request takes the single-rate component of the implicit Adams development associated with [this pull request](https://github.com/inducer/leap/pull/4) and splits it into a separate PR, while also attempting to fix a number of documentation and refactoring issues raised in the initial PR. Among the largest changes are

- Adding a base class for Adams methods to remove duplicate code.
- Adding a number of utilities to that base class to remove duplicate code, largely associated with history rotation.
- Moving the implicit solve used by both RK and Adams methods to `implicit.py`.
- Refactoring the `generate_butcher` function in the RK base class to allow for Adams methods to use this function for bootstrapping.
- Modifying the simple convergence testing utility to allow for implicit testing, and folding Adams-Moulton testing into test_ab.py (`test_ab.py` -> `test_adams.py`).
- Adding a lot of documentation, but I am still iterating on this and trying to get this part exactly right.

inducer commented on leap/multistep/__init__.py:
```diff
@@ -219,7 +222,7 @@ class AdamsBashforthMethodBuilder(MethodBuilder):
 
     def __init__(self, component_id, function_family=None, state_filter_name=None,
-            hist_length=None, static_dt=False, order=None):
+            hist_length=None, static_dt=False, order=None, _extra_bootstrap=False):
```
I'm not liking this `_extra_bootstrap` thing, mostly because it's ill-defined.

- What happens with the extra history after bootstrap?
- What are its semantics in the explicit case?
- The justification for it seems really weak and mostly focused on debugging?

cmikida2 replied:
This is an option that was previously in place to pass a certain test that compared multi-rate IMEX to single-rate implicit. Given that that's not what we're talking about here, this is an artifact of outdated development and needs to be removed. I'm on it.

cmikida2 replied:
This option and its artifacts are removed by ae9d04f.

inducer commented on leap/multistep/__init__.py:
```diff
@@ -385,35 +424,176 @@ def rk_bootstrap(self, cb):
+        rhs_var_to_unknown[rhs_next_var] = unkvar
+
+        # In implicit mode, the time history must
+        # include the *next* point in time.
```
This doesn't make sense to me. If it includes the future, why is it called "history"? It seems that this was done just so that `emit_adams_integration` could be passed the whole history as it is? If so, I don't agree with that way of writing this code.

cmikida2 replied:
I did it this way both for that reason and because it does *eventually* become history after this step, but I can think about a different way of doing this.

cmikida2 replied:
The way of doing this is changed by 4a87ba9, mostly by way of rethinking naming convention to avoid using the term "history" where it's unwarranted, and changing the order in which things are done as a result.

inducer commented on leap/multistep/__init__.py:
```diff
@@ -385,35 +424,176 @@ def rk_bootstrap(self, cb):
+
+        # {{{ emit solve if possible
+
+        if unknowns and len(unknowns) == len(equations):
```
What is the `else` to this if? Why is it OK to just go by here if we can't solve?

cmikida2 replied:
The else - as far as I'm aware - would be the fully explicit case, but at bare minimum in that case the if condition isn't indicative of that.

cmikida2 replied:
Hmm. I realize now, though, that I only have this in the Adams-Moulton class, so explicit isn't a concern. Let me think about contingencies/what should be done if we don't have a square system, or if it's even possible for that to happen with the current setup.

cmikida2 replied:
Addressed by 3803398.

inducer replied:
Non-square systems aren't allowed IMO.

cmikida2 replied:
When you say "not allowed" do you mean they should never happen? I added the elifs (and now an else) to this block to catch instances in which a system is non-square, but I'm not sure if you're saying that that's not sufficient, or if the check for a square system isn't needed at all because we should never *not* have one.

inducer commented on leap/multistep/__init__.py:
```diff
@@ -385,35 +424,176 @@ def rk_bootstrap(self, cb):
+        cb(rhs_next_var, self.eval_rhs(self.t + self.dt, self.state))
+
+        for i in range(1, len(self.history)):
+            if self.extra_bootstrap:
+                save_crit = i+1
+            else:
+                save_crit = i
+
+            with cb.if_(self.step, "==", save_crit):
+                cb(self.history[i], rhs_next_var)
+
+                if not self.static_dt:
+                    cb(self.time_history[i], self.t + self.dt)
```
This code seems like it's an adverse consequence of the awkward redefinition of "history".

cmikida2 replied:
That and the extra bootstrapping. Given that it's a function of those two things, I'll need to think about how those things need to change before addressing this.

cmikida2 replied:
This is addressed along with redefinition of history by 4a87ba9.

cmikida2 commented on leap/rk/__init__.py:
```diff
@@ -539,14 +560,178 @@ class RK5MethodBuilder(SimpleButcherTableauMethodBuilder):
+    recycle_last_stage_coeff_set_names = ()
+
+
+class DIRK2MethodBuilder(ImplicitButcherTableauMethodBuilder):
```
This is addressed in 6d93fee, which adds accuracy tests for these methods to the RK testing suite.

cmikida2 commented on leap/implicit.py:
```diff
@@ -86,3 +86,28 @@ def solver_hook(expr, var, id, **kwargs):
+
+def generate_solve(cb, unknowns, equations, var_to_unknown, guess):
+
+    # got a square system, let's solve
```
This change has been added in ae9d04f.

cmikida2 commented:
I've addressed all of your comments above and made a number of changes to the Adams-Moulton code, and while most of them should explain themselves as direct responses to the comments above, I want to also mention that I had to very slightly twiddle the convergence tolerance for the accuracy tests in order for the implicit extended-history tests to pass. The pass criteria was changed from 0.9 * expected_order to 0.89 * expected_order, as the EOCs for these cases were one or two hundredths places away from the initial target. The one-line code change for this is in 4a87ba9.

I have checked that the addition of extended history has the same effect on EOC relative to the square system as in the case of Adams-Bashforth - in both cases, there's a slight drop in EOC, typically on the order of 0.08-0.1. I'm not overly concerned about this, but happy to discuss further, as I don't think you were all too thrilled with the fudge factor that was being applied to begin with.

inducer commented on leap/multistep/__init__.py:
```diff
@@ -385,35 +423,149 @@ def rk_bootstrap(self, cb):
+                    "than equations")
+        elif len(unknowns) < len(equations):
+            raise ValueError("Adams-Moulton implicit timestep has more equations "
+                    "than unknowns")
```
Is there an `else` to this last condition? If not:
```suggestion
                    "than unknowns")
        else:
            assert False
```

cmikida2 replied:
Addressed in 61277d8.

inducer commented on leap/rk/__init__.py:
```diff
@@ -203,166 +258,157 @@ def generate_butcher(self, stage_coeff_set_names, stage_coeff_sets, rhs_funcs,
-                        cb(my_rhs, last_rhss[name])
-                        make_known(my_rhs)
+        def make_known(v):
+            unknowns.discard(v)
```
Why is this here? Is there ever a scenario where a variable would be added to `unknowns` only to have to be removed later?

cmikida2 replied:
This is something that I pulled in from the existing Butcher generator. Truthfully I'm not sure that such a scenario exists, though based on where this function is called, it looks like it might be possible in the case of stage recycling. Let me run some tests to determine whether this is even needed, because it's a fair question.

cmikida2 replied:
Addressed in 61277d8. I've removed this function and replaced calls to it with adding rhs's to knowns, but I agree with you where we should never need to remove things from unknowns, and getting rid of that has no ill effect that I can see.

inducer commented on leap/rk/__init__.py:
```diff
@@ -203,166 +258,157 @@ def generate_butcher(self, stage_coeff_set_names, stage_coeff_sets, rhs_funcs,
+                            "unknowns than equations")
+                elif len(unknowns) < len(equations):
+                    raise ValueError("Runge-Kutta implicit timestep has more "
+                            "equations than unknowns")
```
else assert False?

cmikida2 replied:
Addressed in 61277d8.

inducer commented on leap/rk/__init__.py:
```diff
@@ -203,166 +258,157 @@ def generate_butcher(self, stage_coeff_set_names, stage_coeff_sets, rhs_funcs,
+
+    # {{{ update state and time
+
+    def generate_butcher_finish(self, cb, stage_coeff_set_names,
+            stage_coeff_sets, rhs_funcs, estimate_coeff_set_names,
+            estimate_coeff_sets, stage_rhs_vars, estimate_vars):
```
Explain the circumstances in which "finish" is invoked.

cmikida2 replied:
The reason that "finish" exists separately from "primary" essentially relates to the use of these functions in bootstrapping the multistep methods. This "finish" is invoked immediately after "primary" by regularly generated RK methods, because one of its primary responsibilities is yielding state and incrementing the time, but when bootstrapping a multistep method, this "finish" is not called, because other pieces of code are responsible for doing those things (not to mention there are instances in which other things need to occur in the "finish" of a bootstrap step for a multistep method pertaining to history updates).

inducer reviewed (commented):
Thanks! Once the wrinkles below are addressed, I think this is in good shape.

cmikida2 commented on leap/multistep/__init__.py:
```diff
@@ -367,6 +326,85 @@ def eval_rhs(self, t, y):
+            dt_factor = self.dt
+            t_end = 1
+
+        return time_data, relv_times, dt_factor, t_end
```
I tried to add this to the docstring (added in 61277d8), but there are two main differences:

- relevant_times subtracts out self.t, whereas time_data does not. I am truthfully not sure how critical this is.
- more importantly, time_data is a list, whereas relevant_times is an array.

cmikida2 commented on leap/multistep/__init__.py:
```diff
@@ -367,6 +326,85 @@ def eval_rhs(self, t, y):
+                               component_id=self.component_id,
+                               time_id="", time=self.t)
+
+    def set_up_time_data(self, cb, new_t):
```
Added in 61277d8, but not sure if it was exactly what you're looking for. I went through the purpose of the function, described the argument, and what it takes in and returns. Happy to amend this docstring if it seems unsuitable.

cmikida2 commented on leap/multistep/__init__.py:
```diff
@@ -367,6 +326,85 @@ def eval_rhs(self, t, y):
+            for i in range(self.hist_length):
+                cb(time_data_var[i], time_data[i] - self.t)
+
+            relv_times = time_data_var
```
Addressed in 61277d8.

cmikida2 commented on leap/multistep/__init__.py:
```diff
@@ -348,7 +306,8 @@ def generate(self):
             cb_bootstrap(self.step, self.step + 1)
-            with cb_bootstrap.if_(self.step, "==", self.hist_length):
+            bootstrap_length = self.hist_length
+            with cb_bootstrap.if_(self.step, "==", bootstrap_length):
```
Yeah, this was removed in 61277d8. An artifact of the old extend-bootstrap thingy that I overlooked.
